vesuvius-c.h

```diff
@@ -820,4 +820,3366 @@ void reset_mesh_origin_to_roi(TriangleMesh *mesh, const RegionOfInterest *roi) {
     }
 }
 
+
+//vesuvius notes:
+// - when passing pointers to a _new function in order to fill out fields in the struct (e.g. mesh_new)
+//   the struct will take ownership of the pointer and the pointer shall be cleaned up in the _free function.
+//   The caller loses ownership of the pointer
+// - index order is in Z Y X order
+// - a 0 / SUCCESS return code indicates success for functions that do NOT return a pointer
+
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <ctype.h>
+#include <stdint.h>
+#include <math.h>
+#include <float.h>
+#include <stdbool.h>
+#include <assert.h>
+
+
+#include <sys/stat.h>
+#include <string.h>
+#include <errno.h>
+#include <curl/curl.h>
+
+
+#ifndef VESUVIUS_HEADER_ONLY
+  #ifndef VESUVIUS_STATIC_LIB
+    #ifndef VESUVIUS_DYNAMIC_LIB
+      #define VESUVIUS_HEADER_ONLY
+    #endif
+  #endif
+#endif
+
+
+#ifdef VESUVIUS_HEADER_ONLY
```

Comment:
I don't object to this, but I am curious as to why it was done this way. What's the benefit of using the __attribute__ over static which gets you... a fair amount of the way, and is sort of the traditional thing to do for a single-header lib?

Usually the way these things are done is everybody includes the header, one file in the project #defines something like VESUVIUS_IMPL 1 and off we go to the races. See Sean Barrett's libraries and commentary.

Reply:
The PUBLIC / PRIVATE #defines are trying to support a couple different ways of using the code:

1) as a static inline header only library where the implementation is compiled into each including .c file
2) as a VESUVIUS_IMPL style define where the implementation is only in one .c file and the header is just #ifdefed as a bunch of prototypes.
3) as a dynamic library, where the public methods all have to be exported by the .c file and the header contains no inline public methods

I didn't quite finish out 2, since I'd still need to do all of the #ifdefing to include/disclude function implementations, but marking everything as `static inline` solves the problem of the user needing to `#define VESUVIUS_IMPLEMENTATION` in a single .c file.

Then we can use just PUBLIC/PRIVATE to hide the linkage and toggle `static inline` annotation as needed.